src/dbnode/storage/bootstrap/bootstrapper/fs/migrator/migrator.go

```diff
@@ -0,0 +1,261 @@
+	metadata       namespace.Metadata
+	shard          uint32
+}
+
+// mergeKey is the unique set of data that identifies an ReadInfoFileResult.
+type mergeKey struct {
+	metadata   namespace.Metadata
+	shard      uint32
+	blockStart int64
+}
+
+// completedMigration is the updated ReadInfoFileSet after a migration has been performed
+// plus the merge key, so that we can properly merge the updated result back into
+// infoFilesByNamespace map.
+type completedMigration struct {
+	key                   mergeKey
+	updatedInfoFileResult fs.ReadInfoFileResult
+}
+
+// Run runs the migrator
+func (m *Migrator) Run(ctx context.Context) error {
+	ctx, span, _ := ctx.StartSampledTraceSpan(tracepoint.BootstrapperFilesystemSourceMigrator)
+	defer span.Finish()
+
+	// Find candidates
+	candidates := m.findMigrationCandidates()
+	if len(candidates) == 0 {
+		m.log.Debug("no filesets to migrate. exiting.")
+		return nil
+	}
+
+	m.log.Info(fmt.Sprintf("found %d filesets to migrate. fileset migration start", len(candidates)))
+
+	nowFn := m.fsOpts.ClockOptions().NowFn()
+	begin := nowFn()
+
+	// Setup workers to perform migrations
+	var (
+		numWorkers = m.migrationOpts.Concurrency()
+		workers    = make([]*worker, 0, numWorkers)
+		outputCh   = make(chan completedMigration, len(candidates))
+	)
+
+	for i := 0; i < numWorkers; i++ {
+		// Give each worker their own persist manager so that we can write files concurrently.
+		pm, err := fs.NewPersistManager(m.fsOpts)
+		if err != nil {
+			return err
+		}
+		worker := &worker{
+			inputCh:        make(chan migrationCandidate, workerChannelSize),
+			outputCh:       outputCh,
+			persistManager: pm,
+			taskOptions: migration.NewTaskOptions().
+				SetFilesystemOptions(m.fsOpts).
+				SetStorageOptions(m.storageOpts),
+		}
+		workers = append(workers, worker)
+	}
+	closedWorkerChannels := false
+	closeWorkerChannels := func() {
+		if closedWorkerChannels {
+			return
+		}
+		closedWorkerChannels = true
+		for _, worker := range workers {
+			close(worker.inputCh)
+		}
+	}
+	// NB(nate): Ensure that channels always get closed.
+	defer closeWorkerChannels()
+
+	// Start up workers. Intentionally not using sync.WaitGroup so we can know when the last worker
+	// is finishing so that we can close the output channel.
+	activeWorkers := atomic.NewUint32(0)
+	for _, worker := range workers {
+		worker := worker
+		activeWorkers.Inc()
+		go func() {
+			m.startWorker(worker)
+			if activeWorkers.Dec() == 0 {
+				close(worker.outputCh)
+			}
+		}()
+	}
+
+	// Start up goroutine consuming worker output
+	var (
+		migrationResults = make(map[mergeKey]fs.ReadInfoFileResult, len(candidates))
+		resultsConsumed  = make(chan bool, 1)
+	)
+	go func() {
+		for result := range outputCh {
+			migrationResults[result.key] = result.updatedInfoFileResult
+		}
+		close(resultsConsumed)
+	}()
+
+	// Enqueue work for workers
+	for i, candidate := range candidates {
+		worker := workers[i%numWorkers]
+		worker.inputCh <- candidate
```

Comment:
I believe `candidates` are not streamed; can probably simplify by removing the `inputCh` on the workers, and instead have something like

```
candidatesPerWorker := len(candidates) / numWorkers
candidateIdx := 0
for i, worker := range workers {
  nextIdx := candidateIdx + candidatesPerWorker
  if i == len(workers) - 1 {
    nextIdx = len(candidates)
  }

  m.startWorker(worker, candidates[candidateIdx:], outputCh)
  candidateIdx = nextIdx
}
```

Reply:
Oh that's clever. I like it!